apps/web/vibes/2px/components/tabs/index.tsx

```diff
@@ -0,0 +1,118 @@
+'use client'
+
+import { CSSProperties, ReactNode, useEffect, useRef, useState } from 'react'
+
+import * as RadixTabs from '@radix-ui/react-tabs'
+import clsx from 'clsx'
+
+type Tab = {
+  title?: string
+  children?: ReactNode
+}
+
+export type Props = {
+  className?: string
+  tabs?: Tab[]
+}
+
+function usePollAnimationFrame(callback: (timestamp: number) => unknown) {
+  useEffect(() => {
+    let requestId: number
+
+    function poll(timestamp: number) {
+      requestId = requestAnimationFrame(poll)
+
+      callback(timestamp)
+    }
+
+    requestId = requestAnimationFrame(poll)
+
+    return () => {
+      cancelAnimationFrame(requestId)
+    }
+  })
+}
+
+const Tabs = ({ className, tabs }: Props) => {
+  const [value, setValue] = useState('0')
+  const container = useRef<HTMLDivElement>(null)
+  const [activeItemLeft, setActiveItemLeft] = useState(0)
+  const [activeItemWidth, setActiveItemWidth] = useState(0)
+
+  usePollAnimationFrame(() => {
+    if (!container.current) return
+
+    const activeItem = container.current.querySelector(`[data-state="active"]`)
+
+    if (!activeItem) return
+
+    const activeItemRect = activeItem.getBoundingClientRect()
+    const containerRect = container.current.getBoundingClientRect()
+
+    setActiveItemLeft(activeItemRect.left - containerRect.left + container.current.scrollLeft)
+    setActiveItemWidth(activeItemRect.width)
+  })
```

Comment:
Something interesting here is that we're calling `setState` and the `useEffect` in `usePollAnimationFrame` doesn't have a dependencies array. This means that these state changes could cause the component to go into an infinite render loop. I checked and it doesn't happen and I think it's because React is de-opting from the update if the state doesn't change to a different value by reference. Since the active item left and width are the same number, there isn't an issue. Maybe we want to set things up so that something like that isn't a problem. An easy way to avoid this is to put the callback in a ref that's set whenever `usePollAnimationFrame` is called and giving the `useEffect` an empty dependency array:

```ts
function usePollAnimationFrame(callback: (timestamp: number) => unknown) {
  const callbackRef = useRef(callback)

  callbackRef.current = callback

  useEffect(() => {
    let requestId: number

    function poll(timestamp: number) {
      requestId = requestAnimationFrame(poll)

      callbackRef.current(timestamp)
    }

    requestId = requestAnimationFrame(poll)

    return () => {
      cancelAnimationFrame(requestId)
    }
  }, [])
}
```

Reply:
This wasn't needed in our case @migueloller, I just removed it. but surely what you proposed is much safer.